klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+    if "antenna" in str(layout_path):
+        switches += " --antenna_only"
+    elif "density" in str(layout_path):
+        switches += " --density_only"
+
+    # Creating run folder structure
+    pattern_clean = ".".join(os.path.basename(layout_path).split(".")[:-1])
+    output_loc = f"{run_dir}/{table_name}"
+    pattern_log = f"{output_loc}/{pattern_clean}_drc.log"
+
+    # command to run drc
+    call_str = f"python3 {drc_dir}/run_drc.py --path={layout_path} {switches} --table={table_name} --run_dir={output_loc} --run_mode=flat --thr=1  > {pattern_log} 2>&1"
+
+    # Starting klayout run
+    os.makedirs(output_loc, exist_ok=True)
+    try:
+        check_call(call_str, shell=True)
+    except Exception as e:
+        pattern_results = glob.glob(os.path.join(output_loc, f"{pattern_clean}*.lyrdb"))
+        if len(pattern_results) < 1:
+            logging.error("%s generated an exception: %s" % (pattern_clean, e))
+            traceback.print_exc()
+            raise
+
+    # Checking if run is completed or failed
+    pattern_results = glob.glob(os.path.join(output_loc, f"{pattern_clean}*.lyrdb"))
+
+    # Get list of rules covered in the test case
+    rules_tested = get_unit_test_coverage(layout_path)
+
+    if len(pattern_results) > 0:
+        # db to gds conversion
+        marker_output, runset_analysis = convert_results_db_to_gds(pattern_results[0], rules_tested)
+
+        # Generating merged testcase for violated rules
+        merged_output = generate_merged_testcase(layout_path, marker_output)
+
+        # Generating final db file
+        if os.path.exists(merged_output):
+            final_report = f'{merged_output.split(".")[0]}_final.lyrdb'
+            call_str = f"klayout -b -r {runset_analysis} -rd input={merged_output} -rd report={final_report}"
+            check_call(call_str, shell=True)
+
+            if os.path.exists(final_report):
+                rule_counts = parse_results_db(final_report)
+                return rule_counts
+            else:
+                return rule_counts
+        else:
+            return rule_counts
+    else:
+        return rule_counts
+
+
+def run_all_test_cases(tc_df, drc_dir, run_dir, num_workers):
+    """
+    This function run all test cases from the input dataframe.
+
+    Parameters
+    ----------
+    tc_df : pd.DataFrame
+        DataFrame that holds all the test cases information for running.
+    drc_dir : string or Path
+        Path string to the location of the drc runsets.
+    run_dir : string or Path
+        Path string to the location of the testing code and output.
+    num_workers : int
+        Number of workers to use for running the regression.
+
+    Returns
+    -------
+    pd.DataFrame
+        A pandas DataFrame with all test cases information post running.
+    """
+
+    results_df_list = []
+    tc_df["run_status"] = "no status"
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
```

Comment:
same comment about using https://pypi.org/project/pytest-xdist/ here with separate pytest test cases here.

Reply:
will be fixed in #8